File: web/packages/engine/src/op.ts
import type { Hlc, Stamp } from "./hlc.js";
import type { ObjectId, OpId } from "./ids.js";
import type { ObjectKind, Value } from "./value.js";
import { validate as validateFrac } from "./fracindex.js";

export const MAX_PROPS_PER_OP = 64;
export const MAX_STRING_BYTES = 32 * 1024;
export const MAX_KEY_BYTES = 64;

export type Entries = ReadonlyArray<readonly [string, Value]>;

export type OpKind =
  | { readonly op: "create"; readonly kind: ObjectKind; readonly props: Entries }
  | { readonly op: "set_props"; readonly object: ObjectId; readonly entries: Entries }
  | { readonly op: "delete"; readonly object: ObjectId }
  | { readonly op: "restore"; readonly object: ObjectId };

export interface Op {
  readonly id: OpId;
  readonly hlc: Hlc;
  readonly kind: OpKind;
}

export function opObject(op: Op): ObjectId {
  return op.kind.op === "create" ? op.id : op.kind.object;
}

export function opStamp(op: Op): Stamp {
  return { hlc: op.hlc, replica: op.id.replica };
}

export function opEntries(op: Op): Entries {
  switch (op.kind.op) {
    case "create":
      return op.kind.props;
    case "set_props":
      return op.kind.entries;
    default:
      return [];
  }
}

const KEY_RE = /^[A-Za-z0-9_.-]{1,64}$/;

export function isValidKey(key: string): boolean {
  return KEY_RE.test(key);
}

const utf8 = new TextEncoder();

/** Shape validation; mirrors `Op::validate` in Rust. Returns an error message or null. */
export function validateOp(op: Op): string | null {
  const entries = opEntries(op);
  if (entries.length > MAX_PROPS_PER_OP) return `too many props (${entries.length})`;
  const seen = new Set<string>();
  for (const [key, value] of entries) {
    if (!isValidKey(key)) return `bad key ${JSON.stringify(key)}`;
    if (seen.has(key)) return `duplicate key ${key}`;
    seen.add(key);
    switch (value.t) {
      case "f64":
        if (!Number.isFinite(value.v)) return `non-finite float for ${key}`;
        break;
      case "str":
        if (utf8.encode(value.v).length > MAX_STRING_BYTES) return `string too long for ${key}`;
        break;
      case "frac":
        if (!validateFrac(value.v)) return `bad fractional index for ${key}`;
        break;
      default:
        break;
    }
  }
  return null;
}

File: crates/converge-core/src/op.rs
use crate::hlc::{Hlc, Stamp};
use crate::ids::{ObjectId, OpId};
use crate::value::{ObjectKind, Value};

pub const MAX_PROPS_PER_OP: usize = 64;
pub const MAX_STRING_BYTES: usize = 32 * 1024;
pub const MAX_KEY_BYTES: usize = 64;

/// One atomic intent from one replica.
#[derive(Clone, PartialEq, Debug)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Op {
    pub id: OpId,
    pub hlc: Hlc,
    pub kind: OpKind,
}

#[derive(Clone, PartialEq, Debug)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(tag = "op", rename_all = "snake_case"))]
pub enum OpKind {
    /// Creates the object whose id is this op's id.
    Create {
        kind: ObjectKind,
        props: Vec<(String, Value)>,
    },
    /// Writes each entry as an LWW register at this op's stamp.
    SetProps {
        object: ObjectId,
        entries: Vec<(String, Value)>,
    },
    Delete {
        object: ObjectId,
    },
    Restore {
        object: ObjectId,
    },
}

#[derive(thiserror::Error, Debug, PartialEq, Eq)]
pub enum ValidationError {
    #[error("too many props in one op ({0} > {MAX_PROPS_PER_OP})")]
    TooManyProps(usize),
    #[error("property key {0:?} is not a valid key")]
    BadKey(String),
    #[error("duplicate property key {0:?} in one op")]
    DuplicateKey(String),
    #[error("non-finite float for key {0:?}")]
    NonFiniteFloat(String),
    #[error("string for key {0:?} exceeds {MAX_STRING_BYTES} bytes")]
    StringTooLong(String),
    #[error("fractional index for key {0:?} is malformed")]
    BadFracIndex(String),
}

/// Keys are short ASCII identifiers so that byte order == code-unit order in
/// every implementation.
pub fn is_valid_key(key: &str) -> bool {
    !key.is_empty()
        && key.len() <= MAX_KEY_BYTES
        && key
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || b == b'_' || b == b'.' || b == b'-')
}

impl Op {
    /// The object this op touches.
    pub fn object(&self) -> ObjectId {
        match &self.kind {
            OpKind::Create { .. } => self.id,
            OpKind::SetProps { object, .. }
            | OpKind::Delete { object }
            | OpKind::Restore { object } => *object,
        }
    }

    pub fn stamp(&self) -> Stamp {
        Stamp::new(self.hlc, self.id.replica)
    }

    /// Shape validation performed by the server before acceptance and by the
    /// client before authoring. Does not depend on document state.
    pub fn validate(&self) -> Result<(), ValidationError> {
        let entries: &[(String, Value)] = match &self.kind {
            OpKind::Create { props, .. } => props,
            OpKind::SetProps { entries, .. } => entries,
            OpKind::Delete { .. } | OpKind::Restore { .. } => &[],
        };
        if entries.len() > MAX_PROPS_PER_OP {
            return Err(ValidationError::TooManyProps(entries.len()));
        }
        for (i, (key, value)) in entries.iter().enumerate() {
            if !is_valid_key(key) {
                return Err(ValidationError::BadKey(key.clone()));
            }
            if entries[..i].iter().any(|(k, _)| k == key) {
                return Err(ValidationError::DuplicateKey(key.clone()));
            }
            match value {
                Value::F64(f) if !f.is_finite() => {
                    return Err(ValidationError::NonFiniteFloat(key.clone()))
                }
                Value::Str(s) if s.len() > MAX_STRING_BYTES => {
                    return Err(ValidationError::StringTooLong(key.clone()))
                }
                Value::FracIndex(s) if crate::fracindex::validate(s).is_err() => {
                    return Err(ValidationError::BadFracIndex(key.clone()))
                }
                _ => {}
            }
        }
        Ok(())
    }
}
